### PR TITLE
[Feature] Spec-Overlap supporting DP-ATTN; PD-Disaggregation; npugraph mode

### PR DESCRIPTION
**Pull request by iforgetmyname**
## Motivation

Since SGLang has already implemented speculative decoding with overlapped scheduler, we would like to enhance such a feature with supporting DP-Attention with MoE models and PD Disaggregation.
Moreover, this PR also introduced NPUGraph support for draft worker v2 that heavily reduce launching overheads.

## Modifications

- Dealing with padding issues and idle batches when dp-attention is enabled
- Pre-alloc enough space supposing all draft tokens will be accpeted under pd disaggregation
- add draft worker npugraph support, extracting abstract functions from cudagraph

## Benchmarking and Profiling

Bechmarking on NPU:
image

Profiling (CUDA profilings are reduced to only 4 layers):
https://drive.google.com/drive/folders/1qRrbQTO-2ia-23N4yevHVKk2Jr2ZrGDX?usp=sharing

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by ShangmingCai** [comment]
So is it ready for review now? Will find some time this week or maybe next week to finish the PD part.

**Comment by ping1jing2** [comment]
> So is it ready for review now? Will find some time this week or maybe next week to finish the PD part.

yes, it's ok, will finish description today, you can review it firstly, thank you.

File: python/sglang/srt/disaggregation/decode.py [hunk 1]
```diff
@@ -1029,7 +1029,13 @@ def get_new_prebuilt_batch(self: Scheduler) -> Optional[ScheduleBatch]:
         # construct fake completed prefill
         new_batch.prepare_for_prebuilt_extend()
         new_batch.process_prebuilt_extend(self.server_args, self.model_config)
-
+        if self.spec_algorithm.is_eagle() and self.enable_overlap:
+            new_batch.spec_info.future_indices = self.future_map.alloc_future_indices(
```
Comment (iforgetmyname) [inline python/sglang/srt/disaggregation/decode.py]: move this assignment into `new_batch.process_prebuilt_extend()`, pass in `self.future_map` as args

File: python/sglang/srt/managers/overlap_utils.py [hunk 2]
```diff
@@ -114,17 +119,42 @@ def resolve_future(self, model_worker_batch: ModelWorkerBatch):
...
+            return start >= stop
+        else:
+            return start <= stop
+
     def store_to_map(
         self, future_indices: FutureIndices, batch_result: GenerationBatchResult
     ):
         intv = future_indices.interval
         if self.spec_algo.is_eagle():
+            if self.is_empty_slice(intv):
```
Comment (iforgetmyname) [inline python/sglang/srt/managers/overlap_utils.py]: extract these logics into a separate func

File: python/sglang/srt/managers/overlap_utils.py [hunk 3]
```diff
@@ -114,17 +119,42 @@ def resolve_future(self, model_worker_batch: ModelWorkerBatch):
...
+                return
             draft_input: EagleDraftInput = batch_result.next_draft_input
             self._lazy_init_buf(draft_input)
-            self.topk_p_buf[intv] = draft_input.topk_p
-            self.topk_index_buf[intv] = draft_input.topk_index
-            self.hidden_states_buf[intv] = draft_input.hidden_states
-            self.verified_id_buf[intv] = draft_input.verified_id
-            self.new_seq_lens_buf[intv] = draft_input.new_seq_lens
+            if self.buf_initialized:
```
Comment (iforgetmyname) [inline python/sglang/srt/managers/overlap_utils.py]: P.S. add comments for these if-elses

File: python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py [hunk 4]
```diff
@@ -165,6 +165,10 @@ def process_prebuilt_extend(
                 topk_index=topk_index,
                 hidden_states=hidden_states,
                 verified_id=self.output_ids,
+                new_seq_lens=self.seq_lens,
+                allocate_lens=self.seq_lens,
+                num_tokens_per_batch=1,
```
Comment (iforgetmyname) [inline python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py]: add comments here

File: python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py [hunk 5]
```diff
@@ -152,7 +158,14 @@ def __init__(self, eagle_worker: EAGLEWorker):
                 vocab_size = self.model_runner.model_config.vocab_size
 
             self.next_token_logits_buffer = torch.zeros(
-                (self.max_bs, vocab_size),
+                (
+                    (
+                        self.max_bs * self.num_tokens_per_bs
+                        if not hasattr(eagle_worker, "model_runner")
```
Comment (iforgetmyname) [inline python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py]: use `self.forward_mode = ForwardMode.DRAFT_EXTEND_V2`

File: python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py [hunk 6]
```diff
@@ -398,15 +425,26 @@ def replay(self, forward_batch: ForwardBatch):
...
 
         # Replay
-        self.graphs[bs].replay()
+        self.raw_bs = raw_bs
+        self.bs = bs
+        self._update_and_replay(forward_batch)
         out = self.output_buffers[bs]
-        if bs != raw_bs:
+
+        if self.forward_mode == ForwardMode.DRAFT_EXTEND_V2:
```
Comment (iforgetmyname) [inline python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py:440]: add comments here to explain why `DRAFT_EXTEND_V2` use `num_tokens` to unpadding

File: python/sglang/srt/speculative/eagle_info.py [hunk 7]
```diff
@@ -751,12 +753,12 @@ def filter_batch(self, new_indices: torch.Tensor, has_been_filtered: bool = True
 
     def merge_batch(self, spec_info: "EagleDraftInput"):
         if self.future_indices is not None:
-            assert spec_info.future_indices is not None
-            self.future_indices = FutureIndices(
-                indices=torch.cat(
-                    [self.future_indices.indices, spec_info.future_indices.indices]
+            if spec_info.future_indices is not None:
```
Comment (iforgetmyname) [inline python/sglang/srt/speculative/eagle_info.py]: revert this

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 8]
```diff
@@ -408,20 +433,23 @@ def _draft_extend_for_prefill(
...
+                )
+                pt += extend_len
 
         # Construct spec_info
         next_draft_input = EagleDraftInput(
             hidden_states=target_hidden_states,
             verified_id=next_token_ids,
             new_seq_lens=batch.seq_lens,
             allocate_lens=batch.seq_lens,
+            num_tokens_per_batch=1,
```
Comment (iforgetmyname) [inline python/sglang/srt/speculative/eagle_worker_v2.py:452]: add comments

File: python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py [hunk 9]
```diff
@@ -101,7 +102,10 @@ def prepare_for_prebuilt_extend(self: ScheduleBatch):
         )
 
     def process_prebuilt_extend(
-        self: ScheduleBatch, server_args: ServerArgs, model_config: ModelConfig
+        self: ScheduleBatch,
+        server_args: ServerArgs,
+        model_config: ModelConfig,
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py]: If `model_config` is not used, please remove this argument.

File: python/sglang/srt/managers/overlap_utils.py [hunk 10]
```diff
@@ -114,17 +119,35 @@ def resolve_future(self, model_worker_batch: ModelWorkerBatch):
...
+    def store_to_map_for_new_batch(
+        self, future_indices: FutureIndices, draft_input: EagleDraftInput
+    ):
+        intv = future_indices.interval
+        if self.spec_algo.is_eagle():
+            # idle indices do not need store info
+            if self.is_empty_slice(intv):
+                return
+            self._lazy_init_buf(draft_input)
+            if self.buf_initialized:
```
Comment (hnyls2002) [inline python/sglang/srt/managers/overlap_utils.py]: Why we need this `if`

File: python/sglang/srt/managers/overlap_utils.py [hunk 11]
```diff
@@ -114,17 +119,35 @@ def resolve_future(self, model_worker_batch: ModelWorkerBatch):
...
+            intv = future_indices.interval
             self.token_ids_buf[intv] = batch_result.next_token_ids
+
+    def store_to_map_for_new_batch(
+        self, future_indices: FutureIndices, draft_input: EagleDraftInput
+    ):
+        intv = future_indices.interval
+        if self.spec_algo.is_eagle():
+            # idle indices do not need store info
+            if self.is_empty_slice(intv):
```
Comment (hnyls2002) [inline python/sglang/srt/managers/overlap_utils.py]: move this out of `if self.spec_algo.is_eagle()` for general purpose.

File: python/sglang/srt/managers/overlap_utils.py [hunk 12]
```diff
@@ -102,7 +102,12 @@ def resolve_future(self, model_worker_batch: ModelWorkerBatch):
         if self.spec_algo.is_eagle():
             # TODO(lsyin): write future indices into spec_info.future_indices
             draft_input: EagleDraftInput = model_worker_batch.spec_info
-            if draft_input is None:
+            if (
```
Comment (hnyls2002) [inline python/sglang/srt/managers/overlap_utils.py]: This if condition appears overly complex and could introduce additional maintenance overhead. If possible, consider simplifying it by using more unified variables.

File: python/sglang/srt/managers/scheduler.py [hunk 13]
```diff
@@ -859,14 +859,19 @@ def init_disaggregation(self):
             )
 
             # The decode requests pending for pre-allocation
+            if self.draft_worker is None or self.spec_algorithm.is_ngram():
```
Comment (hnyls2002) [inline python/sglang/srt/managers/scheduler.py]: Code duplication for `PREFILL` and `DECODE` branches. Please simplify

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 14]
```diff
@@ -157,11 +157,28 @@ def can_run(self, forward_batch: ForwardBatch):
...
+            torch.cuda.synchronize()
+            self.model_runner.tp_group.barrier()
+            run_once_fn()
+
+    def _capture_graph(self, graph, pool, stream, run_once_fn):
+        with torch.cuda.graph(graph, pool=pool, stream=stream):
+            out = run_once_fn()
+        return out
+
+    def _update_and_replay(self, forward_batch: ForwardBatch):
```
Comment (hnyls2002) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: curious why `_update_and_replay` instead of `_replay`, update happens elsewhere actually.

File: python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py (hunk 9 above)
Comment (iforgetmyname) [reply python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py]: removed

File: python/sglang/srt/managers/overlap_utils.py (hunk 10 above)
Comment (iforgetmyname) [reply python/sglang/srt/managers/overlap_utils.py]: a redundant condition check confirmed, `self.is_empty_slice` should already cover this case

File: python/sglang/srt/managers/overlap_utils.py (hunk 11 above)
Comment (iforgetmyname) [reply python/sglang/srt/managers/overlap_utils.py]: done

File: python/sglang/srt/managers/scheduler.py (hunk 13 above)
Comment (iforgetmyname) [reply python/sglang/srt/managers/scheduler.py]: done

File: python/sglang/srt/managers/overlap_utils.py (hunk 12 above)
Comment (iforgetmyname) [reply python/sglang/srt/managers/overlap_utils.py]: leftover codes when debugging thus removed. when we are processing idle batch, `draft_input` should always be none

**Review by ShangmingCai — APPROVED** [review]
The diff of the disaggregation part is clean and LGTM.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py (hunk 14 above)
Comment (iforgetmyname) [reply python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: to clearify, npu graph runner does require an extra update of a cpu tensor here to determine the tiling of the attention op. however the naming is still confusing from the view of cuda graph runner, we have changed it to `_replay`

**Review by hnyls2002 — APPROVED** [review]

**Comment by iforgetmyname** [comment]
> CI is broken https://github.com/sgl-project/sglang/actions/runs/19293114439/job/55183519002?pr=12443#step:5:3666

I see, I copied this test case from `sglang/test/srt/test_eagle_dp_attention.py` and just found that some of the specs are not supported by spec-overlap now. shoud pick up another successful case and modify it to support dp-attention, any suggestions on such a case?
